File: frontend/chatwmex_app_frontend/lib/models/voice_message.dart
class VoiceMessage {
  final String id;
  final String senderId;
  final String senderName;
  final String roomId;
  final String fileUrl;
  final int duration; // 语音时长（秒）
  final int fileSize; // 文件大小（字节）
  final DateTime timestamp;
  final MessageType type;

  VoiceMessage({
    required this.id,
    required this.senderId,
    required this.senderName,
    required this.roomId,
    required this.fileUrl,
    required this.duration,
    required this.fileSize,
    required this.timestamp,
    this.type = MessageType.voice,
  });

  factory VoiceMessage.fromJson(Map<String, dynamic> json) {
    return VoiceMessage(
      id: json['id'] ?? '',
      senderId: json['sender_id'] ?? '',
      senderName: json['sender_name'] ?? '',
      roomId: json['room'] ?? '',
      fileUrl: json['file_url'] ?? '',
      duration: json['duration'] ?? 0,
      fileSize: json['file_size'] ?? 0,
      timestamp: DateTime.parse(json['timestamp'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'sender_id': senderId,
      'sender_name': senderName,
      'room': roomId,
      'file_url': fileUrl,
      'duration': duration,
      'file_size': fileSize,
      'timestamp': timestamp.toIso8601String(),
      'type': 'voice',
    };
  }
}

// 扩展现有的 MessageType 枚举
enum MessageType {
  text,
  image,
  file,
  system,
  voice, // 新增语音类型
}

File: backend/models/voice_message.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// VoiceMessage 代表一條語音消息
type VoiceMessage struct {
	ID         primitive.ObjectID `bson:"_id,omitempty" json:"id"`
	SenderID   string             `bson:"sender_id" json:"sender_id"`
	SenderName string             `bson:"sender_name" json:"sender_name"`
	Room       string             `bson:"room" json:"room"`
	FilePath   string             `bson:"file_path" json:"file_path"` // 加密後的文件路徑
	Duration   int                `bson:"duration" json:"duration"`   // 語音時長（秒）
	FileSize   int64              `bson:"file_size" json:"file_size"` // 文件大小（字節）
	Timestamp  time.Time          `bson:"timestamp" json:"timestamp"`
	Type       string             `bson:"type" json:"type"` // 消息類型，固定為 "voice"
}